File: 10. Dynamic Programming/07. Best time to buy and sell stock iv/main.py
# 1. Brute Force (Recursive) - Exponential Time Complexity (O(2^N))
def brute_force(prices, i, transactions_left, holding):
    if i >= len(prices) or transactions_left == 0:
        return 0

    if holding:
        return max(prices[i] + brute_force(prices, i + 1, transactions_left - 1, False),
                   brute_force(prices, i + 1, transactions_left, True))
    else:
        return max(-prices[i] + brute_force(prices, i + 1, transactions_left, True),
                   brute_force(prices, i + 1, transactions_left, False))

def max_profit_brute_force(k, prices):
    return brute_force(prices, 0, k, False)


# 2. Recursion + Memoization (O(N * K) Time, O(N * K) Space)
def memoized_dfs(prices, i, transactions_left, holding, memo):
    if i >= len(prices) or transactions_left == 0:
        return 0
    if memo[i][transactions_left][holding] != -1:
        return memo[i][transactions_left][holding]

    if holding:
        memo[i][transactions_left][1] = max(prices[i] + memoized_dfs(prices, i + 1, transactions_left - 1, False, memo),
                                            memoized_dfs(prices, i + 1, transactions_left, True, memo))
    else:
        memo[i][transactions_left][0] = max(-prices[i] + memoized_dfs(prices, i + 1, transactions_left, True, memo),
                                            memoized_dfs(prices, i + 1, transactions_left, False, memo))

    return memo[i][transactions_left][holding]

def max_profit_memo(k, prices):
    n = len(prices)
    memo = [[[-1] * 2 for _ in range(k + 1)] for _ in range(n)]
    return memoized_dfs(prices, 0, k, False, memo)


# 3. Bottom-Up Dynamic Programming (O(N * K) Time, O(N * K) Space)
def max_profit_dp(k, prices):
    n = len(prices)
    if n == 0:
        return 0

    dp = [[0] * n for _ in range(k + 1)]

    for t in range(1, k + 1):
        max_diff = -prices[0]
        for i in range(1, n):
            dp[t][i] = max(dp[t][i - 1], prices[i] + max_diff)
            max_diff = max(max_diff, dp[t - 1][i] - prices[i])

    return dp[k][n - 1]


# 4. Optimized DP with O(N) Space (O(N * K) Time, O(K) Space)
def max_profit_optimized_dp(k, prices):
    n = len(prices)
    if n == 0:
        return 0

    # When k is large, reduce to infinite transactions problem (Greedy)
    if k >= n // 2:
        profit = sum(max(prices[i] - prices[i - 1], 0) for i in range(1, n))
        return profit

    buy = [-float('inf')] * (k + 1)
    sell = [0] * (k + 1)

    for price in prices:
        for t in range(1, k + 1):
            buy[t] = max(buy[t], sell[t - 1] - price)
            sell[t] = max(sell[t], buy[t] + price)

    return sell[k]


# Main function to test different approaches
def main():
    prices = [3, 2, 6, 5, 0, 3]  # Example input
    k = 2  # Maximum number of transactions

    print("Brute Force:", max_profit_brute_force(k, prices))
    print("Memoization:", max_profit_memo(k, prices))
    print("Bottom-Up DP:", max_profit_dp(k, prices))
    print("Optimized DP:", max_profit_optimized_dp(k, prices))

# Run the main function
main()

File: 10. Dynamic Programming/07. Best time to buy and sell stock iv/main.cpp
#include <iostream>
#include <vector>
#include <climits>

using namespace std;

// 1. Brute Force (Recursive) - Exponential Time Complexity (O(2^N))
int bruteForce(vector<int>& prices, int i, int transactionsLeft, bool holding) {
    if (i >= prices.size() || transactionsLeft == 0) return 0;

    if (holding) {
        return max(prices[i] + bruteForce(prices, i + 1, transactionsLeft - 1, false),
                   bruteForce(prices, i + 1, transactionsLeft, true));
    } else {
        return max(-prices[i] + bruteForce(prices, i + 1, transactionsLeft, true),
                   bruteForce(prices, i + 1, transactionsLeft, false));
    }
}

int maxProfitBruteForce(int k, vector<int>& prices) {
    return bruteForce(prices, 0, k, false);
}

// 2. Recursion + Memoization (O(N * K) Time, O(N * K) Space)
int memoizedDFS(vector<int>& prices, int i, int transactionsLeft, bool holding, vector<vector<vector<int>>>& memo) {
    if (i >= prices.size() || transactionsLeft == 0) return 0;
    if (memo[i][transactionsLeft][holding] != -1) return memo[i][transactionsLeft][holding];

    if (holding) {
        memo[i][transactionsLeft][1] = max(prices[i] + memoizedDFS(prices, i + 1, transactionsLeft - 1, false, memo),
                                           memoizedDFS(prices, i + 1, transactionsLeft, true, memo));
    } else {
        memo[i][transactionsLeft][0] = max(-prices[i] + memoizedDFS(prices, i + 1, transactionsLeft, true, memo),
                                           memoizedDFS(prices, i + 1, transactionsLeft, false, memo));
    }

    return memo[i][transactionsLeft][holding];
}

int maxProfitMemo(int k, vector<int>& prices) {
    int n = prices.size();
    vector<vector<vector<int>>> memo(n, vector<vector<int>>(k + 1, vector<int>(2, -1)));
    return memoizedDFS(prices, 0, k, false, memo);
}

// 3. Bottom-Up Dynamic Programming (O(N * K) Time, O(N * K) Space)
int maxProfitDP(int k, vector<int>& prices) {
    int n = prices.size();
    if (n == 0) return 0;

    vector<vector<int>> dp(k + 1, vector<int>(n, 0));

    for (int t = 1; t <= k; t++) {
        int maxDiff = -prices[0];
        for (int i = 1; i < n; i++) {
            dp[t][i] = max(dp[t][i - 1], prices[i] + maxDiff);
            maxDiff = max(maxDiff, dp[t - 1][i] - prices[i]);
        }
    }
    return dp[k][n - 1];
}

// 4. Optimized DP with O(N) Space (O(N * K) Time, O(K) Space)
int maxProfitOptimizedDP(int k, vector<int>& prices) {
    int n = prices.size();
    if (n == 0) return 0;

    if (k >= n / 2) { // When k is large, reduce to infinite transactions problem (Greedy)
        int profit = 0;
        for (int i = 1; i < n; i++) {
            if (prices[i] > prices[i - 1])
                profit += prices[i] - prices[i - 1];
        }
        return profit;
    }

    vector<int> buy(k + 1, INT_MIN);
    vector<int> sell(k + 1, 0);

    for (int price : prices) {
        for (int t = 1; t <= k; t++) {
            buy[t] = max(buy[t], sell[t - 1] - price);
            sell[t] = max(sell[t], buy[t] + price);
        }
    }
    return sell[k];
}

// Main function to test different approaches
int main() {
    vector<int> prices = {3, 2, 6, 5, 0, 3}; // Example input
    int k = 2; // Maximum number of transactions

    cout << "Brute Force: " << maxProfitBruteForce(k, prices) << endl;
    cout << "Memoization: " << maxProfitMemo(k, prices) << endl;
    cout << "Bottom-Up DP: " << maxProfitDP(k, prices) << endl;
    cout << "Optimized DP: " << maxProfitOptimizedDP(k, prices) << endl;

    return 0;
}
